Код ревью библиотеки

SeriousMeow opened the pull request.

amethyst1 committed:
Настроить окружение

SeriousMeow committed:
Настрой .clang-format и .clang-tidy

SeriousMeow committed:
Удали Cmake настройки для динамических библиотек и Windows

SeriousMeow committed:
Добавь библиотеку GLM

SeriousMeow committed:
Добавь генерацию документации

SeriousMeow committed:
Добавь графические примитивы

SeriousMeow committed:
Добавь 3D объект

SeriousMeow committed:
Удали ненужные директории

SeriousMeow committed:
Добавь сцену

SeriousMeow committed:
Добавь камеру

SeriousMeow committed:
Добавь изображение

SeriousMeow committed:
Исправь документацию

SeriousMeow committed:
Добавь конструктор объекта и расширь документацию

SeriousMeow committed:
Расширь интерфейс сцены и добавь документацию

SeriousMeow committed:
Добавь рендерер

SeriousMeow committed:
Добавь недостающие методы и изменения

SeriousMeow committed:
Добавь методы создания матриц трансформации и исправь ошибки использования библиотеки

SeriousMeow committed:
Добавь загрузку obj файлов

SeriousMeow committed:
Исправь ошибки отрисовки и исправь документацию

SeriousMeow committed:
Улучши код рендерера

SeriousMeow committed:
Добавь примеры использования

SeriousMeow committed:
Измени базовые типы для удобства

В для базовых типов точки и вектора теперь используются декартовы
координаты, что облегчает работу с ними

SeriousMeow committed:
Исправь опечатку в лицензии

SeriousMeow committed:
Улучши интерфейс изображения

Добавлены методы взаимодействия с пикселем по заданным координатам

SeriousMeow committed:
Расширь интерфейс сцены и камеры

Добавлены новые методы взаимодействия со сценой и камерой. Теперь
возможно менять положение камеры и положение объектов на сцене. Также
названия типов изменены на более конкретные

SeriousMeow committed:
Исправь опечатки

SeriousMeow committed:
Исправь ошибки в документации

DimaTrushin commented on example/BMP_Writer/BMP_Writer.hpp:
```diff
@@ -0,0 +1,85 @@
+    Pixel* const GetPixel(const int32_t x, const int32_t y);
+
+private:
+#pragma pack(push, 1)
```
Нахрена, а главное зачем? Тут много вопросов
1) Почему не
```cpp
#pragma pack(push, 2)
```
? Ибо у тебя тут идут данные размеров в 2 и 4 байта. И выравнивать по 1 байту глупо.
2) Что мешало переставить местами file_size и offset_data в начало? Тогда ты будешь иметь только потерю 2 байтов в конце.
3) Если же потеря двух байт в конце критична, то все равно можно использовать pack и перестановку полей структуры, чтобы у тебя не нарушалось выравнивание для 4 байтного инта.
4) Ты вообще тестировал эффект от этого безобразия?
5) Я бы делал такие вещи отключаемыми одним макосом. То есть я бы сделал какой-то служебный макрос
```cpp
#define PACK_CLASS_BEGIN ...
#define PACK_CLASS_END
```
и потом бы использовал
```cpp
PACK_CLASS_BEGIN
struct Data {
  ...
};
PACK_CLASS_END
```
И сделал бы макрос, который бы включал или отключал эту запаковку для всех типов.

DimaTrushin replied:
Все специфическое для файлов я бы вынес в отдельный класс
```cpp
class BMPReader{
  ...
};
```
И тогда бы пользовался так
```cpp
BMPReader r;
Image img = read.from("file.bmp");
// либо
Image img;
r >> img;
```
И аналогично для записи.

DimaTrushin commented on example/BMP_Writer/BMP_Writer.hpp:
```diff
@@ -0,0 +1,85 @@
+    };
+#pragma pack(pop)
+
+#pragma pack(push, 1)
```
Здесь упаковка ни на что не влияет. У тебя и так все выравнено в памяти и заполнено без padding-ов.

DimaTrushin commented on example/BMP_Writer/BMP_Writer.hpp:
```diff
@@ -0,0 +1,85 @@
+namespace bmp {
+
+namespace bmp_constants {
+const uint16_t kFileType = 0x4D42;
```
Вот это у тебя ODR Violation. Ты объявляешь глобальную переменную в хедере. Если этот хедер будет включен в несколько единиц трансяции, то это нарушение ODR. Так как у тебя встроенные типы, просто напиши constexpr вместо const.

DimaTrushin commented on example/BMP_Writer/BMP_Writer.hpp:
```diff
@@ -0,0 +1,85 @@
+    double r;
+    double g;
+    double b;
+    Pixel& operator+=(const Pixel& rhs);
+    Pixel operator*(double value) const;
+    void Normalization();
```
Не делай методы для структур. Это не очень хорошо. Структуры лучше использовать для хранения данных, а мтеоды делать глобальными функциями или статическими.

DimaTrushin commented on example/BMP_Writer/BMP_Writer.hpp:
```diff
@@ -0,0 +1,85 @@
+
+class Image {
+public:
+    Image(const int32_t width, const int32_t height);
```
Сырые типы в интерфейсе конструктора. Нет возможности при вызове понять это смысл этих констант. Например ты можешь написать
```cpp
Image img(100, 200);
```
И хрен его знает ты их правильно задал или перепутал высоту и ширину. Можно использовать strongly typed aliases следующим образом
```cpp
enum Width : int23_t;
enum Height : int32_t;
// и потом использовать
Image img(Width{100}, Height{200});
```
И компилятор ругнется, если ты туда подставишь просто целое число или перепутаешь их местами. Это стандартная техника.

DimaTrushin commented on example/BMP_Writer/BMP_Writer.hpp:
```diff
@@ -0,0 +1,85 @@
+class Image {
+public:
+    Image(const int32_t width, const int32_t height);
+    void Save(const std::string& save_path);
```
Плохая идея в класс пихать все подряд. Главная задача методов класса -- обеспечить поддержание инвариантов класса. Плохо, чтобы класс умел себя писать куда-то. Для этого твоему классу надо знать кучу иррелевантных для его работы деталей и он становится зависимым от различных источников, в которые ты хочешь писать. Правильно сделать метод сериализации в абстрактный поток и тогда ты можешь писать и в строку, и в файл, и в буффер в памяти и в сеть и куда угодно не меняя код.

DimaTrushin commented on example/BMP_Writer/BMP_Writer.hpp:
```diff
@@ -0,0 +1,85 @@
+    const int32_t GetHeight() const;
+
+    const Pixel* const GetPixel(const int32_t x, const int32_t y) const;
+
```
Ты не консистентен с пробелами между функциями. Почему-то выше у теб доступ к ширине и высоте без пробела, а тут одна и та же версия функции перегруженная по this внезапно разделены строкой.

DimaTrushin commented on example/BMP_Writer/BMP_Writer.hpp:
```diff
@@ -0,0 +1,85 @@
+    const int32_t GetWidth() const;
+    const int32_t GetHeight() const;
+
+    const Pixel* const GetPixel(const int32_t x, const int32_t y) const;
```
А почему возвращается указатель на пиксель? Почему не ссылка? Он может быть nullptr?

DimaTrushin commented on example/BMP_Writer/BMP_Writer.hpp:
```diff
@@ -0,0 +1,85 @@
+
+    const Pixel* const GetPixel(const int32_t x, const int32_t y) const;
+
+    Pixel* const GetPixel(const int32_t x, const int32_t y);
```
Судя по всему тут ты хочешь получать доступ к пикселям и менять их. Но тогда тут плохо две вещи:
1) Почему возвращается указатель, а не ссылка?
2) Плохое название метода. Имя Get обычно говорить "возьми данное из класса" и такой метод обычно не влияет на состояние класса.

DimaTrushin commented on example/BMP_Writer/BMP_Writer.hpp:
```diff
@@ -0,0 +1,85 @@
+    const uint32_t PixelToBits(const Pixel& pixel) const;
+    BMPFileHeader file_header_;
+    BMPInfoHeader bmp_info_header_;
+    std::vector<std::vector<Pixel>> image_;
```
Никогда НЕ используй вектор векторов. используй один вектор и храни картинку нарезанную например на строки в одном большом векторе. Это всегда лучше по кэш мисам.

DimaTrushin commented on example/BMP_Writer/BMP_Writer.hpp:
```diff
@@ -0,0 +1,85 @@
+    void CheckFileHeader() const;
+    void CheckInfoHeader() const;
+    void GenerateHeaders();
+    const Pixel BitsToPixel(const uint32_t bits) const;
+    const uint32_t PixelToBits(const Pixel& pixel) const;
```
Добавь пустую строку между данными и фукнциями класса.

DimaTrushin replied:
Вот это выглядит как вспомогательные функции для пикселя и не ясно почему эти методы находятся тут в классе Image.

DimaTrushin commented on example/BMP_Writer/BMP_Writer.hpp:
```diff
@@ -0,0 +1,85 @@
+        uint32_t colors_important{0};
+    };
+#pragma pack(pop)
+    void CheckFileHeader() const;
+    void CheckInfoHeader() const;
+    void GenerateHeaders();
```
Ну вот у тебя image занимается не своим делом. Он начинает зависеть от знания чего-то профайлы и их структуру.

DimaTrushin commented on example/BMP_Writer/BMP_Writer.cpp:
```diff
@@ -0,0 +1,159 @@
+#include <cassert>
+#include <fstream>
+
+bmp::Pixel& bmp::Pixel::operator+=(const Pixel& rhs) {
```
Не делай длинные квалификаторы доступа пиши так
```cpp
namespace bmp {
Pixel& Pixel::operator+=(const Pixel& rhs) {
  ...
}
}
```
Получится намного более читаемо

DimaTrushin commented on example/BMP_Writer/BMP_Writer.cpp:
```diff
@@ -0,0 +1,159 @@
+}
+
+void bmp::Pixel::Normalization() {
+    const double zero{0};
+    const double one{1};
```
Ну так уж можно не делать. Имена этих констант ничего не добавляют к информации. И кроме того, почему не constexpr?

DimaTrushin commented on example/BMP_Writer/BMP_Writer.cpp:
```diff
@@ -0,0 +1,159 @@
+        assert((width > 0) and "Image: ширина должна быть больше 0");
+        assert((heiht > 0) and "Image: высота должна быть больше 0");
+    }
+    bmp_info_header_.width = width;
+    bmp_info_header_.height = heiht;
+    image_ = std::vector<std::vector<bmp::Pixel>>(heiht, std::vector<bmp::Pixel>(width));
```
Питонячий стиль. Это должно быть сделано в списке инициализации. Если ты хочешь вставить assert до списка инициализации, надо в списке инициализации вставлять метод очистки следующим образом
```cpp
class Image {
  Image(int32_t width, int32_t height)
    : bmp_info_header_(fill_header(width, height)),
      image(width * height) {}
  static BMPInfoHeader fill_header(int32_t width, int32_t height) {
    assert(Условие && "комментарий");
    return ....;
  }
};
```

DimaTrushin commented on example/BMP_Writer/BMP_Writer.cpp:
```diff
@@ -0,0 +1,159 @@
+        throw std::runtime_error("Unable to open the output image file");
+    }
+    GenerateHeaders();
+    image.write(reinterpret_cast<const char*>(&file_header_), sizeof(file_header_));
```
reinterpret_cast должне быть частью очень низкоуровневого кода, а не в бизнес логике.

DimaTrushin commented on example/BMP_Writer/BMP_Writer.hpp:
```diff
@@ -0,0 +1,85 @@
+    Image(const int32_t width, const int32_t height);
+    void Save(const std::string& save_path);
+
+    const int32_t GetWidth() const;
```
нет смысла возвращать результат из функции по константному значению.

DimaTrushin commented on example/example_load_obj/example_load_obj.cpp:
```diff
@@ -0,0 +1,44 @@
+
+#include "../BMP_Writer/BMP_Writer.hpp"
+
+int main() {
```
У тебя не отключены exception, потому надо поставить блок на них. Это делается так
```cpp
#include "Application.h"
#include "Except.h"

int main() {
  try {
    renderer::Application app();
    app.run();
  } catch(...) {
    except::react();
  }
  return 0;
}
Except.h
namespace except {
void react();
}
Except.cpp
#include "Except.h"

namespace except {
void react() {
  try {
    throw;
  } catch(std::exception& e) {
    // обрабатываешь известные исключения
  } catch(...) {
    // обрабатываешь незивестные исключения
  }
}
}
```
Функция main никогда не выполняет конкретную работу. Она должна заниматься оркестровкой и делегирует все другому коду. Кроме того, она является барьером для утекания ошибок и исключений в ОС. Не надо путать ошибки которые возникли в приложении из-за приложения с ошибками, которые возникли в приложении из-за ОС. И не надо засорять ОС ошибками.

DimaTrushin reviewed (changes requested):
Сделал несколько общих замечаний на сейчас.

SeriousMeow committed:
Измени расположение заголовков

Заголовочные файлы библиотеки были перенесены из директории include в
директорию source

SeriousMeow committed:
Вынеси глобальные типы в отдельный заголовок

SeriousMeow committed:
Переименуй тип TransformMatrix

Тип renderer::TransformMatrix переименован в renderer::Matrix

SeriousMeow committed:
Добавь функционал разработчика в документацию

Добавлена возможность генерировать дополнительные разделы документации в
режиме разработчика

SeriousMeow committed:
Измени архитектуру сцены

SeriousMeow committed:
Добавь защиту на размеры изображения

Добавлена защита на размеры изображения. Также удалена ненужная
переменная

SeriousMeow committed:
Исправь константность методов изображения

Метод GetPixel у renderer::Image теперь константный

SeriousMeow committed:
Улучши вспомогательный функционал для записи изображений

Удалена система с bmp::Image. Теперь существует единственная фукция
SaveToBmp для записи renderer::Image в BMP файл

SeriousMeow committed:
Измени название функций сцены

Renderer::Scene::IsCameraExists и Renderer::Scene::IsObjectExists переименованы в
Renderer::Scene::HasCamera и Renderer::Scene::HasObject

SeriousMeow committed:
Удали ненужный код

Удален ненужный код в Renderer::Image

SeriousMeow committed:
Добавь проверки корректности при изменении пикселя

SeriousMeow committed:
Улучши читаемость кода

DimaTrushin commented on example/BMP_Writer/BMP_Writer.hpp:
```diff
@@ -0,0 +1,22 @@
+ * @bug Может не работать для систем с big endian
+ * @todo Сделать универсальную работу с endianess
+ */
+void SaveToBmp(const renderer::Image& image, const std::string& path);
```
Никогда не пиши ничего в global namespace. Заведи единое пространство имен для проекта и пиши в него. Если тебе нужен библиотечный код, то заведи для него отдельное пространство имен.

DimaTrushin commented on example/BMP_Writer/BMP_Writer.cpp:
```diff
@@ -0,0 +1,111 @@
+const uint32_t kBlueMask = 0x0000FF;
+const uint32_t kBitsInByte = 8;
+const uint32_t kCompressionTypeUncopressed = 0;
+}  // namespace
```
Не закрывай пространство имен после каждого чиха. Если оно должно тянуться непрерывно, то пиши в нем как в непрерывном scope. Так будет проще читать и следить за кодом и где он находится.

DimaTrushin commented on example/BMP_Writer/BMP_Writer.cpp:
```diff
@@ -0,0 +1,111 @@
+ * @brief Константы для создания BMP файла
+ */
+namespace {
+const uint16_t kFileType = 0x4D42;
```
constexpr

DimaTrushin commented on example/BMP_Writer/BMP_Writer.cpp:
```diff
@@ -0,0 +1,111 @@
+
+void SaveToBmp(const renderer::Image& image, const std::string& path) {
+    {
+        assert((image.GetWidth() > 0) and "SaveToBmp: ширина должна быть больше 0");
```
В целом плохая идея писать сообщения об ошибках на русском или вообще любом другом языке отличном от английского. Могут быть проблемы с отображением шрифтов в разных системах. Но НЕ надо это править. Пусть будет так.

DimaTrushin commented on example/BMP_Writer/BMP_Writer.cpp:
```diff
@@ -0,0 +1,111 @@
+    }
+    std::ofstream ouput_image{path, std::ios_base::binary};
+    if (not ouput_image) {
+        throw std::runtime_error("SaveToBmp: не удалось открыть указанный файл для записи");
```
Смотри, бросание исключения уронит твою программу, если ты его не перехватишь, что не есть хорошее поведение. Ну и что, что ты не смог записать в файл, это не повод твоей программе умирать. Потому лучше использовать return и возвращать код ошибки. Ты все равно возвращаешь void и по сути не используешь return. Возвращай статус выполненной операции.

DimaTrushin commented on example/BMP_Writer/BMP_Writer.cpp:
```diff
@@ -0,0 +1,111 @@
+    for (int32_t row = 0; row < info_header.height; ++row) {
+        for (int32_t column = 0; column < info_header.width; ++column) {
+            const renderer::Image::Pixel current_pixel =
+                image.GetPixel(column, info_header.height - 1 - row);  // BMP пишется снизу вверх
```
Вот это пример хорошего комментария.

DimaTrushin commented on example/example_load_obj/example_load_obj.cpp:
```diff
@@ -0,0 +1,31 @@
+    renderer::Renderer renderer;
+
+    // Параметры
+    const size_t width = 1280;
```
constexpr

DimaTrushin commented on example/example_load_obj/example_load_obj.cpp:
```diff
@@ -0,0 +1,31 @@
+    image = renderer.Render(scene, camera_id, std::move(image));
+
+    // Записываем результат в файл
+    SaveToBmp(image, "example_load_obj.bmp");
```
Вот это должно быть в utils

DimaTrushin commented on source/renderer/utils.cpp:
```diff
@@ -0,0 +1,43 @@
+Object LoadObjFile(const std::string& path) {
+    std::ifstream input{path};
+    if (not input.is_open()) {
+        throw std::runtime_error{"LoadObjFile: не удалось открыть файл"};
```
Не лучшая политика кидать исключения и валить программу на лопатки при ошибки чтения из файла. Я бы сделал так. Я бы обернул эту функцию дополнительую оболочку так
```cpp
utils.h
std::optional<Object> LoadObjFile(const std::string& path);
utils.cpp
namespace {
Object LoadObjFileHelper(const std::string& path) {
  /*содержание текущей функции*/
}
}
std::optional<Object> LoadObjFile(const std::string& path) {
  try {
    return LoadObjeFileHelper(path);
  } catch (...) {
    return std::nullopt;
  }
}
```
Так ты ставишь границу, где перехватываются исключения. Это не глобальные ошибки. Ты можешь разобраться с ними в точке вызова функции чтения из файла. Во время чтения из файла действительно удобно кидать исключения, потому что если пошла ошибка в какой-то внутренней мелкой функции парсинга, тебе надо просто завершить весь парсинг и выйти из всех вызовов вспомогательных функций. И исключения дают удобный способ это сделать. Однако, в точке вызова чтения из файла ты всего лишь хочешь получить информацию, что файл не был прочитан из-за какой-то ошибки и получить пустой результат.

DimaTrushin commented on source/renderer/utils.cpp:
```diff
@@ -0,0 +1,43 @@
+            triangles.push_back(new_triangle);
+        }
+    }
+    input.close();
```
Тебе не надо закрывать файловые потоки явно, их закрывает деструктор. Не делай работу, которая будет выполнена автоматом. Для этого и надо RAII.

DimaTrushin commented on source/renderer/utils.cpp:
```diff
@@ -0,0 +1,43 @@
+        std::stringstream s;
+        s << line;
+        if (line[0] == 'v') {
+            char _;
+            Point new_point;
+            s >> _ >> new_point.x >> new_point.y >> new_point.z;
+            points.push_back(new_point);
```
Вот это должна быть одна строчка
```cpp
points.push_back(extract_point(s));
```

DimaTrushin commented on source/renderer/utils.cpp:
```diff
@@ -0,0 +1,43 @@
+            points.push_back(new_point);
+        }
+        if (line[0] == 'f') {
+            char _;
+            Triangle new_triangle;
+            size_t index;
+            s >> _;
+            for (size_t i = 0; i < 3; ++i) {
+                s >> index;
+                new_triangle.vertices[i].point = points[index - 1];
+            }
+            triangles.push_back(new_triangle);
```
Вот это должна быть одна строчка
```cpp
triangles.push_back(extract_triangle(s));
```

DimaTrushin commented on source/renderer/utils.cpp:
```diff
@@ -0,0 +1,43 @@
+        }
+        std::stringstream s;
+        s << line;
+        if (line[0] == 'v') {
```
Вот такое условие лучше завернуть в функцию с говорящим названием. Например
```cpp
if (is_point_next(line))
```
Или что-то в этом духе. Так будет проще читать код и понимать, что по смыслу за условие проверяется, а не просто низкоуровненвую проверку битов.

DimaTrushin commented on source/renderer/types.hpp:
```diff
@@ -0,0 +1,51 @@
+ * Матрицы 4x4 используются для выстроения связей между объектами в пространстве. Рекомендуется
+ * создавать только с помощью функций библиотеки или использовать объявленные в библиотеки константы
+ */
+using Matrix = glm::mat4;
```
А матрицы 3 на 3 у тебя нигде не используются?

DimaTrushin commented on source/renderer/transform.hpp:
```diff
@@ -0,0 +1,150 @@
+ * Не производит никаких преобразований. Совмещает начала и направления координатных осей без
+ * изменения масштаба
+ */
+const Matrix kNoTransforms{1.0f};
```
1) ODR violation. Ты нарушаешь правило одного определения. Это UB. Дело в том, что если этот хедер будет включен в две единицы трансляции, то компилятор в каждой из них сделает по глобальной переменной с таким именем. И линкер увидет две переменные с одинаковым именем. Объявления глобальных переменных и функций в h файлах надо снабжать inline спецификатором.

2) В целом плохая тема использовать глобальные константы. Если они тебе понадобятся в конструкторах глобальных объектов, то у тебя могут быть ошибки инициализации связанные с линковкой, которые очень тяжело найти. Дело в том, что между разными единицами трансляции порядок инициализации глобальных объектов не фиксирован. Это просто невозможно сделать. И если ты где-то объявишь другую глобальную константу, которая определена в терминах этой константы, то может быть проблема.

3) Если есть возможность лучше делать constexpr константы. Тогда это избавляет тебя от всех проблем. Но не все библиотеки поддерживают constexpr конструкторы. Это печально обычно.

4) Другой выход -- вместо того, чтобы говорить в терминах матриц -- данных задающих преобразования. Можно говорить в терминах самих преобразований. То есть объявить функцию
```cpp
auto identity(auto&& expr) {
  return std::forward<decltype(expr)>(expr);
}
```
И тогда тебе не надо возиться с глобальными константами.

5) Ну или можешь объявить функцию, которая возвращает единичную матрицу с названием identity.

DimaTrushin commented on source/renderer/transform.hpp:
```diff
@@ -0,0 +1,150 @@
+ *
+ * @return Матрица преобразования
+ */
+Matrix Translate(const Vector& direction);
```
Названия этого метода и всех методов ниже врут. Ты пишешь так, как будто эти функции выполняют преобразования, но они лишь возвращают матрицу, которая выполняет нужное преобразование. Например тут лучше было бы TranslationMatrix и это бы давало понимание, что ты делаешь.

DimaTrushin commented on source/renderer/transform.hpp:
```diff
@@ -0,0 +1,150 @@
+ *
+ * @return Матрица преобразования
+ */
+Matrix CameraLookAtPoint(const Point& camera_position, const Point& center);
```
Я еще понимаю, что предыдущие функции делали абстрактные операции линейной алгебры. А тут камера-специфичное преобразование. Почему эта функция вообще тут? Это часть кода камеры. Одно дело, когда ты пишешь код строящий матрицы имеющие математический смысл, другое дело, когда ты пишешь код специфичный для класса. Например для камеры тебе надо построить некую матрицу. Тогда с точки зрения камеры эта матрица может иметь свое название а-ля local coordinates matrix. Но с точки зрения математики -- это будет orthogonal matrix или что-то в этом духе. Или если тебе нужна матрица проекции, то абстрактно это oprojector matrix, а для камеры это будет screen projection matrix. Так вот камера специфичные методы имплементируются через математические методы. И тут в этом файле должны быть чисто математические методы. А все что связано с конкретными классами должно быть отправлено туда, где эти специфические классы живут.

DimaTrushin commented on source/renderer/transform.cpp:
```diff
@@ -0,0 +1,47 @@
+    return RotateDegrees(angle, Vector{0, 0, 1});
+}
+
+Matrix CameraLookAtPoint(const Point& camera_position, const Point& center) {
```
Я бы ожидал использоваться
```cpp
camera.lookAtPointMatrix(point);
```
И не надо эти методы писать тут. Пусть это будут члены класса камеры. Либо это должны быть приватные методы камеры. Потому что пользователю камеры плевать на матрицы. Ибо для пользователя камеры имеют смысл более высокоуровневые вещи. А именно, пользователь хочет сказать "камера, смотрю на эту точку". И ему плевать как это имплементируется, есть ли там какие-то матрицы и т.д. Потому я бы ожидал, что этот метод это будет какой-то внутренний приватный метод, а камера имеет метод
```cpp
camera.lookAtPoint(point);
```

DimaTrushin commented on source/renderer/scene.hpp:
```diff
@@ -0,0 +1,175 @@
+
+private:
+    std::vector<SceneObject> objects_;
+    std::vector<Matrix> cameras_;
```
Странно, что список камер -- это одна матрица. В этом случае у тебя интерфейс камеры -- это интерфейс матрицы. Кроме того, вдруг кроме этой матрицы у тебя в камере еще что-то будет? Вдруг ты захочешь поменять имплементацию камеры? У тебя сломается весь код. использующий камеру, потому что весь этот код думает, что камера -- это всего лишь матрица. Кроме того, так вообще не понятно что эта матрица значит. Это матрица проекции? Или это матрица локальной системы координат, или что это вообще такое?

DimaTrushin commented on source/renderer/scene.hpp:
```diff
@@ -0,0 +1,175 @@
+        /**
+         * @brief ID объекта
+         */
+        ObjectId id;
```
Зачем это нужно? Ты хочешь адресоваться не по индексу вектора? Просто искать по такому id -- это за линию. А по индексу вектора за O(1).

DimaTrushin commented on source/renderer/scene.hpp:
```diff
@@ -0,0 +1,175 @@
+    /**
+     * @brief Итератор объектов в сцене
+     */
+    using ObjectsIterator = std::vector<SceneObject>::const_iterator;
```
ConstObjectIterator

DimaTrushin commented on source/renderer/scene.hpp:
```diff
@@ -0,0 +1,175 @@
+    /**
+     * @brief Создание пустой сцены
+     */
+    Scene() = default;
```
Зачем тебе объявлять дефолтный конструктор, когда нет никаких других? Он же и так будет сформирован неявно.

DimaTrushin commented on source/renderer/scene.hpp:
```diff
@@ -0,0 +1,175 @@
+     *
+     * @return ID добавленного объекта
+     */
+    ObjectId PushObject(const Object& object, const Matrix& matrix);
```
1) Добавь версию
```cpp
ObjectId PushObject(Object&& object, const Matrix& matrix);
```
Для поглощения объекта, если он дорогой для копирования.
2) Название matrix не отражает назначение матрицы. Какой у нее смысл? Кстати, а какие у нее размеры? Добавь в название Matrix -> Matrix4, чтобы было понятно, что они статического размера и у них тяжелая локальная часть.
3) Плохая идея возвращать Id. Почему не вернуть итератор? как делают все нормальные методы.

DimaTrushin commented on source/renderer/scene.cpp:
```diff
@@ -0,0 +1,60 @@
+
+Scene::ObjectId Scene::PushObject(const Object& object, const Matrix& matrix) {
+    ObjectId id = objects_.size();
+    objects_.push_back({.id = id, .object = object, .object_to_scene = matrix});
```
Нахрена индекс запоминать внутри вектора? Это дублированная информация. Плюс она может быть в рассинхроне.

DimaTrushin commented on source/renderer/scene.cpp:
```diff
@@ -0,0 +1,60 @@
+}
+
+bool Scene::HasCamera(const CameraId id) const {
+    return (0 <= id and id < objects_.size());
```
баг копипасты! objects_ -> cameras_

DimaTrushin commented on source/renderer/scene.cpp:
```diff
@@ -0,0 +1,60 @@
+}
+
+Matrix Scene::GetObjectMatrix(const ObjectId id) const {
+    {
+        assert(HasObject(id) and "GetObjectMatrix: объекта с переданным ID не существует");
+    }
```
А зачем ты так выделяешь ассерт в отдельный scope?

DimaTrushin commented on source/renderer/renderer.hpp:
```diff
@@ -0,0 +1,79 @@
+     *
+     * @return Срендеренное изображение
+     */
+    Image&& Render(const Scene& scene, const Scene::CameraId camera_id, Image&& image,
```
Блять. Убрать Image&& и сделать Image. Это UB возвращать ссылку на локальную переменную.

DimaTrushin commented on source/renderer/renderer.hpp:
```diff
@@ -0,0 +1,79 @@
+     * @return Срендеренное изображение
+     */
+    Image&& Render(const Scene& scene, const Scene::CameraId camera_id, Image&& image,
+                   const float fov_x = 90.0, const float focal_length = 1.0);
```
Волшебные константы. Заведи static constexpr константы с говорящими именами и используй их тут вместо 90.0 и 1.0.

DimaTrushin replied:
Вот эта информация должна быть внутри камеры.

DimaTrushin commented on source/renderer/renderer.hpp:
```diff
@@ -0,0 +1,79 @@
+     * @param[in] focal_length Расстояние от камеры до экрана, должно быть больше 0.0
+     * @param[in] fov_x Горизонтальное поле зрения в градусах, должно быть больше 0.0 и меньше 360.0
+     */
+    void Init(const size_t width, const size_t height, const float focal_length, const float fov_x);
```
Мы обсуждали, что этот метод либо врет своим названием, либо он делает работу конструктора. Ни то ни другое не является поводом его оставить. Более того, последние два аргумента -- это настройки камеры, а не рендерера. Чтобы понять что к чему относится -- подумай, что если ты поменяешь камеру, а что если ты поменяешь рендерер, какие параметры должны сохраняться, а какие поменяться.

DimaTrushin commented on source/renderer/renderer.hpp:
```diff
@@ -0,0 +1,79 @@
+     * @param[in] start Начало отрезка
+     * @param[in] end Конец отрезка
+     */
+    void DrawLine(Image& image, const Point& start, const Point& end);
```
Если у тебя у функции есть in и out параметры, то есть правила, как такие функции писать.
1) Все in параметры должны идти перед out параметрами.
2) Все in параметры передаются или by value или by const reference.
3) Все out параметры передаются by pointer.
Это стандарт даже в Си.

Плюс вызывающий метод видит на стороне вызова, какой аргумент поменяется. Сравни два вызова
```cpp
DrawLine(image, point1, point2);
DrawLine(point1, point2, &image);
```

DimaTrushin commented on source/renderer/renderer.hpp:
```diff
@@ -0,0 +1,79 @@
+    void DrawLine(Image& image, const Point& start, const Point& end);
+
+    struct Parameters {
+        size_t width{0};
+        size_t height{0};
```
Мы обсуджали, что эти параметры -- это часть z буффера. Сделай zBuffer отдельным классом с нужным интерфейсом.

DimaTrushin commented on source/renderer/renderer.hpp:
```diff
@@ -0,0 +1,79 @@
+    struct Parameters {
+        size_t width{0};
+        size_t height{0};
+        float fov_x{0};
+        float focal_length{0};
+        float x_scale{0};
+        float y_scale{0};
+        Matrix camera_to_clip;
```
Это все параметры камеры.

DimaTrushin commented on source/renderer/renderer.cpp:
```diff
@@ -0,0 +1,144 @@
+
+    const uint32_t steps = std::max(std::abs(x_start - x_end), std::abs(y_start - y_end)) + 1;
+
+    const Vector dp = (end - start) / static_cast<float>(steps);
```
Наличие кастов означает, что твои классы передают друг другу информацию не в том формате, в котором ее ожидают другие классы. В данном случае у тебя на плоскости есть вде системы координат: 1) непрерывная 2) дискретная. И тебе нужны методы пересчета этих координат. Скорее всего это методы камеры. И они уже внутри могут тягать касты. Но не тут в бизнес логике.

DimaTrushin commented on source/renderer/renderer.cpp:
```diff
@@ -0,0 +1,144 @@
+            DrawLine(image, Point{vertices[2]}, Point{vertices[0]});
+        }
+    }
+    return std::move(image);
```
Не надо делать std::move из локальных переменных в return. Ты в лучшем случае сломаешь rvo. В худшем UB. Если бы ты написал
```cpp
return image;
```
То компилятор или static analyser кинул бы тебе warning, что ты возвращаешь ссылку на локальный объект. А ты этим кастом не дал компилятору или анализатору это увидеть.

Кроме того, все функции по умолчанию свои локальные переменные мувают в ретерн, если он возвращается by value. Но компилятор может сделать больше. Он может сделать не только move, но и rvo, то есть создать локальный объект по адресу возвращаемого. Но для этого надо чтобы тип возвращаемого объекта совпадал с типом в сигнатуре функции. Если функция имеет вид
```cpp
Image make() {
  Image image;
  return std::move(image);
}
```
То возвращаемый тип функции Image, а выражение в return имеет тип Image&&. И компилятор видит, что типы разные и не может сделать RVO. А если же написано
```cpp
Image make() {
  Image image;
  return image;
}
```
То компилятор просто создаст image по адресу возвращаемого значения и тут не будет ни лишних копирований, ни лишних мувов.

DimaTrushin commented on source/renderer/renderer.cpp:
```diff
@@ -0,0 +1,144 @@
+            Point4 vertices[3];
+            // нормализуем координаты и переводим в координаты камеры
+            for (size_t i = 0; i < 3; ++i) {
+                vertices[i] = Point4{triangle.vertices[i].point, 1.0f};
```
1) 4 вложенных scope. Даже в ядре Linux запрещено иметь более 3. Это всегда ошибка дизайна. Обычно это означает, что ты не объявил нужные вспомогательные функции.
2) Три вложенных цикла друг в друга. Это очень сложно проверять на корректность. Очень полезно внутри цикла напрямую вызывать функцию
```cpp
for (int i = 0; i < size; ++i) {
  x = do_work(data1, data2);
}
```
При таком подходе у тебя появляются четкие границы между внешностью и внутренностью цикла, где ты видишь, что именно подается внутрь тела цикла, и что именно возвращается оттуда. Кроме того, имя функции говорит тебе явно, что за работа делается внутри цикла. В целом вложенные циклы -- это знак того, что ты что-то делаешь не так.
3) Бывают ситуации, когда для имплементации математических алгоритмов приходится писать вложенные циклы и делать более 3 вложенных scope, но это не тот случай.

DimaTrushin commented on source/renderer/renderer.cpp:
```diff
@@ -0,0 +1,144 @@
+
+        assert((scene.HasCamera(camera_id)) and "Render: камера должна принадлежать сцене");
+    }
+    if (image.GetWidth() != parameters_.width or image.GetHeight() != parameters_.height or
+        fov_x != parameters_.fov_x or focal_length != parameters_.focal_length) {
+        Init(image.GetWidth(), image.GetHeight(), focal_length, fov_x);
+    }
```
Если я правильно понимаю это должно быть
```cpp
reallocate_if_needed(width, height, &image);
```
Это нарушение уровня абстракции, ты делаешь руками низкоуровневую работую.

DimaTrushin commented on source/renderer/renderer.cpp:
```diff
@@ -0,0 +1,144 @@
+        fov_x != parameters_.fov_x or focal_length != parameters_.focal_length) {
+        Init(image.GetWidth(), image.GetHeight(), focal_length, fov_x);
+    }
+    z_buffer_.assign(parameters_.width * parameters_.height,
+                     std::numeric_limits<float>::infinity());
```
Заверни это в объект.

DimaTrushin commented on source/renderer/renderer.cpp:
```diff
@@ -0,0 +1,144 @@
+    }
+    z_buffer_.assign(parameters_.width * parameters_.height,
+                     std::numeric_limits<float>::infinity());
+    for (Scene::ObjectsIterator objects_it = scene.ObjectsBegin(); objects_it != scene.ObjectsEnd();
```
Тут достаточно auto, чтобы не мучиться с длинными именами.

DimaTrushin commented on source/renderer/renderer.cpp:
```diff
@@ -0,0 +1,144 @@
+    for (Scene::ObjectsIterator objects_it = scene.ObjectsBegin(); objects_it != scene.ObjectsEnd();
+         ++objects_it) {
+        Matrix object_to_camera = scene.GetCameraMatrix(camera_id) * (*objects_it).object_to_scene;
+        const Object& object = (*objects_it).object;
```
Вот это знак того, что тут должна быть вызвана фукнция вида
```cpp
?? do_something(const Object& object, ...);
```
куда ты передаешь первый аргумент.

DimaTrushin replied:
Почему не написать object_it->object???

DimaTrushin commented on source/renderer/renderer.cpp:
```diff
@@ -0,0 +1,144 @@
+                }
+                vertices[i] /= vertices[i].w;
+            }
+            // находим нормаль к грани
+            Point4 triangle_center = (vertices[0] + vertices[1] + vertices[2]) / 3.0f;
+            Vector triangle_normal =
+                glm::cross(Vector{vertices[1] - vertices[0]}, Vector{vertices[2] - vertices[0]});
```
Вместо комментария, который может устареть и быть неактуальным и который не понимает компилятор и не носитель русского, что мешает сделать так
```cpp
Vector triangle_normal = getNormals(...);
```

DimaTrushin commented on source/renderer/renderer.cpp:
```diff
@@ -0,0 +1,144 @@
+            Vector triangle_normal =
+                glm::cross(Vector{vertices[1] - vertices[0]}, Vector{vertices[2] - vertices[0]});
+
+            // проверяем, что грань направлена в сторону камеры
+            Vector camera_direction = -Vector{triangle_center};
+            if (glm::dot(triangle_normal, camera_direction) < 0.0f) {
+                continue;
+            }
```
Что мешало сделать функцию у камеры с нормальным названием и тут написать так
```cpp
if (!camera.is_front_to_me(triangle)) {
  continue;
}
```
Кроме того, если бы ты тут вызвал функцию обработки треугольника, то можно было бы написать return, а не continue и все зависимости были бы локализованы.

DimaTrushin commented on source/renderer/renderer.cpp:
```diff
@@ -0,0 +1,144 @@
+            if (glm::dot(triangle_normal, camera_direction) < 0.0f) {
+                continue;
+            }
+            for (size_t i = 0; i < 3; ++i) {
+                vertices[i] = parameters_.camera_to_clip * vertices[i];
+                assert((glm::epsilonNotEqual(vertices[i].w, 0.0f, glm::epsilon<float>())) and
+                       "Renderer: после умножения на матрицу перхода в Clip пространство "
+                       "координата w стала 0");
+                vertices[i] /= vertices[i].w;
+            }
```
Судя по всему это одна строчка кода вызова какой-то вспомогательной функции.

DimaTrushin commented on source/renderer/renderer.cpp:
```diff
@@ -0,0 +1,144 @@
+                       "координата w стала 0");
+                vertices[i] /= vertices[i].w;
+            }
+            DrawLine(image, Point{vertices[0]}, Point{vertices[1]});
+            DrawLine(image, Point{vertices[1]}, Point{vertices[2]});
+            DrawLine(image, Point{vertices[2]}, Point{vertices[0]});
```
Вот тут ты вызываешь внешние функции, а весь код выше ты делаешь более мелкую работу. Это называется нарушение уровня абстракции. Если ты хотя бы в одной строчке кода делегируешь работу другой функции. Значит все другие строчки, тоже должны делегировать, а не делать это руками в цикле.

DimaTrushin commented on source/renderer/primitives.hpp:
```diff
@@ -0,0 +1,38 @@
+ *
+ * Вершина в пространстве
+ */
+struct Vertex {
```
А точно удобно зашивать математику в структуры? ты же просто ломаешь интерфейс точки и теперь ее нельзя складывать с вектором и куча других неудобств.

DimaTrushin commented on source/renderer/object.hpp:
```diff
@@ -0,0 +1,61 @@
+     *
+     * @param[in] triangles треугольники
+     */
+    explicit Object(const std::vector<Triangle>& triangles);
```
Плохо раскрывать детали имплементации в конструкторе. Например, ты можешь передать в конструктор любой другой контейнер и если он конвертируется к std::vector, то у тебя произойдет неявная конвертация с копированием данных. Потом ты скопируешь из временного вектора внутрь и у тебя будет 2 лишних копирования тяжелых данных.
Если уж принимать std::vector и раскрывать детали имплементации. То либо принимай его by value  и мувай его во внутренности класса, либо принимай по rvalue ссылке и мувай внутрь.
В любом случае я бы оставил вариант от rvalue ссылки.

DimaTrushin commented on source/renderer/object.hpp:
```diff
@@ -0,0 +1,61 @@
+     *
+     * return Итератор начала контейнера
+     */
+    Iterator Begin();
```
Если уж у тебя есть только один вид итераторов, то делай begin, end и имена iterator и const_iterator совместитмыми с stl контейнерами. Тогда ты сможешь пользоваться stl алгоритмами.

DimaTrushin commented on source/renderer/image.hpp:
```diff
@@ -0,0 +1,118 @@
+     *
+     * @param[in] other Изображение для копирования
+     */
+    Image& operator=(const Image& other);
```
Копирующий оператор тебе перегружать не надо. Он работает и так корректно. Напиши = default.

DimaTrushin commented on source/renderer/image.hpp:
```diff
@@ -0,0 +1,118 @@
+     *
+     * @param[in] other Изображение для копирования
+     */
+    Image(const Image& other);
```
Копирующий конструктор перегружать не надо, он работает корректно. Напиши = default.

DimaTrushin commented on source/renderer/image.hpp:
```diff
@@ -0,0 +1,118 @@
+     *
+     * Каждая компонента цвета имеет диапазон от 0.0 до 1.0
+     */
+    struct Pixel {
```
А точно эта структура должна быть внутри Image, а не отдельно?

DimaTrushin commented on source/renderer/image.cpp:
```diff
@@ -0,0 +1,58 @@
+        assert((width > 0) and "Image: ширина должна быть больше 0");
+        assert((height > 0) and "Image: высота должна быть больше 0");
+    }
+    image_.resize(width_ * static_cast<size_t>(height));
```
Здесь static_cast не нужен. Так как у тебя должна произойти неявная конвертация типов.

DimaTrushin commented on source/renderer/image.cpp:
```diff
@@ -0,0 +1,58 @@
+}
+
+Image& renderer::Image::operator=(Image&& other) {
+    if (this != &other) {
+        width_ = std::move(other.width_);
+        image_ = std::move(other.image_);
+    }
```
Чтобы не делать проверку на самоприсваивание обычно пишут так
```cpp
Image& renderer::Image::operator=(Image&& other) noexcept {
  Image tmp = std::move(other);
  std::swap(tmp. width_, width_);
  std::swap(tmp.image_, image_);
  return *this;
}
```
Так ты манипуляцию ресурсов отдаешь конструкторам и деструкторам. И избавляешься от одного лишнего if-а.

DimaTrushin commented on source/renderer/image.hpp:
```diff
@@ -0,0 +1,118 @@
+     *
+     * @param[in] other Изображение для перемещения
+     */
+    Image(Image&& other);
```
Мув конструктор должен быть обязательно noexcept

DimaTrushin commented on source/renderer/image.hpp:
```diff
@@ -0,0 +1,118 @@
+     *
+     * @param[in] other Изображение для перемещения
+     */
+    Image& operator=(Image&& other);
```
Мув присваивание должно быть обязательно noexcept

DimaTrushin commented on source/renderer/image.cpp:
```diff
@@ -0,0 +1,58 @@
+Image::Image(const Image& other) : width_{other.width_}, image_{other.image_} {
+}
+
+Image::Image(Image&& other) : width_{std::move(other.width_)}, image_{std::move(other.image_)} {
```
1) На width_ нет смысла навешивать std::move. 
2) Это НЕ корректная имплементация. Тебе надо использовать std::exchange. Дело в том, что для встроенных типов move это то же самое что копирование и ты оставил образ в некорректном состоянии после мува. Должно быть
```cpp
Image::Image(Image&& other)
  : width_{std::exchange(other.width_, 0)},
    image_{std::move(other.image_)} {}
```
Для вектора можно оставить как есть. Но можно написать
```cpp
image_{std::exchange(other.image_, {})}
```

DimaTrushin commented on source/renderer/image.cpp:
```diff
@@ -0,0 +1,58 @@
+}
+
+size_t Image::GetHeight() const {
+    return image_.size() / width_;
```
Так как ты разрешил мувать из Image, то у тебя может быть width_ == 0. Потому надо бы тут либо обработать этот случай. Либо добавить хотя бы assert. Но я бы обрабатывал, потому что раз пустое изображение для тебя корректное состояние, то странно, что такие методы как взятие размера могут упасть.

DimaTrushin commented on example/example_simple_piramide/example_simple_piramide.cpp:
```diff
@@ -0,0 +1,47 @@
+#include <renderer/include_all.hpp>
+
+#include "../BMP_Writer/BMP_Writer.hpp"
+
```
В cpp файле можно писать using namespace renderer. Это можно написать в начале int main()

DimaTrushin reviewed (changes requested):
Обрати внимание, что я не повторял замечания. Если одна и та же ошибка всречалась несколько раз, я комментировал лишь единожды. Потому я ожидаю, что ты поправишь все аналогичные замечания всюду.

SeriousMeow committed:
Исправь опечатки

SeriousMeow committed:
Замени const на constexpr

SeriousMeow committed:
Исправь ошибку возвращаемого значения Render

SeriousMeow committed:
Добавь камеры

SeriousMeow committed:
Исправь ошибку развертывания Cmake с включеным dev mode

SeriousMeow committed:
Исправь внутренний интерфейс Renderer

SeriousMeow committed:
Приведи Cmake файлы в единый стиль и исправь ошибку генерации документации

SeriousMeow committed:
Измени интерфейс Image

1. Добавлена поддержка изображений с нулевыми размерами
2. Изменен формат пикселя
3. Удалены геттеры и сеттеры для пикелей, добавлено получение пикселя по
ссылке

SeriousMeow committed:
Исправь отсутствие имплементации части новых методов

SeriousMeow committed:
Исправь неправильную обработку пустого изображения у Renderer

SeriousMeow committed:
Исправь опечатки в документации

DimaTrushin commented:
Гори в аду с кучей своих ревью. Я не понимаю, что я только что отревьювил? Как блять понять, что из этого твой проект, а что случайная копия?

SeriousMeow committed:
Измени систему управления камерой

Теперь камера управляется углами взгляда, в связи с чем удалены функции
создания матриц камеры

Исправлена несогласованность координат координат камеры

SeriousMeow committed:
Добавь функции управления камерой

Добавлены функции для удобного управления камерой: изменение углов
добавлением и сдвиг вдоль направлений относительно направления взгляда

SeriousMeow committed:
Измени систему объектов

Теперь грани объектов в сцене хранятся в одном хранилище, а информация
об объекте хранится в SceneObject. Также этот объект предоставляет
интерфейс для изменения параметров объекта

Удалены функции создания матриц из-за ненадобности

SeriousMeow committed:
Добавь отсечение по пирамиде зрения

Теперь треугольники, выходящие за пирамиду зрения, отсекаются. Это
исправило проблемы падения программы и потери производительности при
отворачивании камеры от объектов и нахождении камеры внутри них

SeriousMeow committed:
Добавь растеризацию граней

SeriousMeow committed:
Исправь опечатки в типе

SeriousMeow committed:
Добавь библиотеку Assimp для импорта моделей

SeriousMeow committed:
Подготовь алгоритм к добавлению многопоточности, освещения и текстур

SeriousMeow committed:
Добавь многопоточную растеризацию

SeriousMeow committed:
Добавь освещение

SeriousMeow committed:
Добавь текстуры и исправь баги

SeriousMeow committed:
Измени материал по-умолчанию на библиотечный

SeriousMeow committed:
Запрети копирование и перемещение Singleton классов

SeriousMeow committed:
Добавь публикацию документации

SeriousMeow committed:
Удали m.css и обнови версию Doxygen

SeriousMeow committed:
Исправь Github Actions

SeriousMeow committed:
Измени Actions для публикации

SeriousMeow committed:
Добавь разрешения

SeriousMeow committed:
Обнови README

SeriousMeow committed:
Удали BUILDING.md

SeriousMeow committed:
Убери библиотеку stb_image.h из исходного кода

SeriousMeow committed:
Исправь вычисление освещения
